### PR TITLE
write simulation details to log file.

### PR DESCRIPTION
**Pull request by qilicun**
Write the terminal output messages to log file, needs opm-core https://github.com/OPM/opm-core/pull/1010.

### CODE REVIEW

File: opm/autodiff/FlowMain.hpp [hunk 1]
```diff
@@ -677,15 +677,20 @@ namespace Opm
 
             if (!schedule->initOnly()) {
                 if (output_cout_) {
-                    std::cout << "\n\n================ Starting main simulation loop ===============\n"
-                              << std::flush;
+                    std::string msg;
+                    msg = "\n\n================ Starting main simulation loop ===============\n";
+                    std::cout << msg << std::endl;
```
Comment (joakim-hove) [inline opm/autodiff/FlowMain.hpp]: See: https://github.com/OPM/opm-core/pull/1010#discussion_r62454356

File: opm/autodiff/BlackoilModelBase_impl.hpp [hunk 2]
```diff
@@ -282,8 +283,10 @@ namespace detail {
                 current_relaxation_ -= nonlinear_solver.relaxIncrement();
                 current_relaxation_ = std::max(current_relaxation_, nonlinear_solver.relaxMax());
                 if (terminalOutputEnabled()) {
-                    std::cout << " Oscillating behavior detected: Relaxation set to "
-                              << current_relaxation_ << std::endl;
+                    std::string msg = " Oscillating behavior detected: Relaxation set to "
+                        + std::to_string(current_relaxation_);
+                    std::cout << msg << std::endl;
```
Comment (atgeirr) [inline opm/autodiff/BlackoilModelBase_impl.hpp]: This should not be necessary, the info() call below should be sufficient to print this to screen if that has been requested (and by default it should).

File: opm/autodiff/BlackoilModelBase_impl.hpp [hunk 3]
```diff
@@ -1106,6 +1109,7 @@ namespace detail {
         if (converged) {
             if ( terminal_output_ ) {
                 std::cout << "well converged iter: " << it << std::endl;
+                OpmLog::info("well converged iter: " + std::to_string(it));
```
Comment (atgeirr) [inline opm/autodiff/BlackoilModelBase_impl.hpp]: Same problem here as above: using OpmLog should be sufficient.

**Comment by atgeirr** [comment]
I checked and see that given the current logging setup in FlowMain.hpp we do not get log output to cout. I think it is relatively easy to change though: add a StreamLog (or EclipsePRTLog) backend that writes to cout in FlowMain.hpp.

**Comment by qilicun** [comment]
That can be solved by OPM/opm-core#1010 (comment). There is a problem(maybe not a problem but an issues)here, in OpmLog, we have some convenient methods ``info(), error()``,which record information  with prefixed messages, the screen output will like "Info: xxxxx", "Warning:", this is useful in ``relpermDiagnostics``, but for flow looks strange. One solution for this will be drop the prefix, let the user add when it's needed.

**Comment by atgeirr** [comment]
> but for flow looks strange

I agree, we should consider dropping the prefix altogether. An idea: what about color coding? (debug, info -> black; warning -> yellow; problem, error, bug -> red)

**Comment by qilicun** [comment]
> An idea: what about color coding? (debug, info -> black; warning -> yellow; problem, error, bug -> red)

Using color is a good idea, but this should be achieved by using prefix message? Otherwise, maybe macro?

**Comment by atgeirr** [comment]
> Using color is a good idea, but this should be achieved by using prefix message? Otherwise, maybe macro?

What about adding a flag "use_color_coding_" to StreamLog, and if that is true we'll add terminal color codes before streaming the message in addMessage()?

**Comment by atgeirr** [comment]
Thinking a little more about it: it seems to me we should generalize slightly how we mark messages of different types in the output: prefix, color, both or nothing. This could be handled in StreamLog (which should be called StreamLogBackend) or even the LogBackend base class.

**Comment by qilicun** [comment]
>What about adding a flag "use_color_coding_" to StreamLog, and if that is true we'll add terminal color codes before streaming the message in addMessage()?

Sounds good! I think ``addMessage`` in ``LogBackend`` should be the place, because if you use color, that means you set prefix, the user should not prefix messages by themselves, otherwise it will contain prefix twice.

**Comment by atgeirr** [comment]
Yes, the current implementation adds the prefix in OpmLog::warning(), OpmLog::info() etc. That should be removed, and the back-end should handle it based on configuration.

**Comment by atgeirr** [comment]
Actually, addMessage() in LogBackend is not quite the place, since that is the method that is overridden by the subclasses (it should even be made pure virtual). Rather, add a protected method LogBackend::decorateMessage() that modifies the message string based on the type and configuration (such as use_prefix_, use_color_coding_ etc.), and call that from the addMessage() methods of subclasses.

**Comment by qilicun** [comment]
>Yes, the current implementation adds the prefix in OpmLog::warning(), OpmLog::info() etc. That should be removed, and the back-end should handle it based on configuration.

Since you think a lot of this, would you mind to take that task?

**Comment by atgeirr** [comment]
I'll make a proposal to opm-common.

**Comment by qilicun** [comment]
>I'll make a proposal to opm-common.

Great! After that, I will update this PR and apply some changes in ``relpermDiagnostics``

**Comment by atgeirr** [comment]
You could experiment with OPM/opm-common#115.

File: opm/autodiff/FlowMain.hpp [hunk 4]
```diff
@@ -264,9 +264,9 @@ namespace Opm
                 std::cout << "**********************************************************************\n";
                 std::cout << "*                                                                    *\n";
                 std::cout << "*                   This is Flow (version " << version << ")"
-                          << std::string(26 - version.size(), ' ') << "*\n";
```
Comment (atgeirr) [inline opm/autodiff/FlowMain.hpp]: There seems to be some wrong indentation introduced here and below, please fix.

**Comment by atgeirr** [comment]
On the whole I very much want to merge this, but I am reluctant to do so before we have at least a workaround for the problem that terminal output is now buried by the relperm diagnostics (i.e. a max count of some sort as discussed in OPM/opm-core#1011).

**Comment by joakim-hove** [comment]
> but I am reluctant to do so before we have at least a workaround for the problem that terminal output is now buried by the relperm diagnostics

Possible approach: https://github.com/OPM/opm-core/pull/1011#issuecomment-219203707?

**Comment by qilicun** [comment]
Depend on https://github.com/OPM/opm-common/pull/119/

**Comment by atgeirr** [comment]
Thanks, will merge.

**Pull request merged**